src/librustdoc/passes/check_custom_code_classes.rs

```diff
@@ -0,0 +1,92 @@
+//! NIGHTLY & UNSTABLE CHECK: custom_code_classes_in_docs
+//!
+//! This pass will produce errors when finding custom classes outside of
+//! nightly + relvant feature active.
+
+use super::{span_of_attrs, Pass};
+use crate::clean::{Crate, Item};
+use crate::core::DocContext;
+use crate::fold::DocFolder;
+use crate::html::markdown::{find_testable_code, ErrorCodes, LangString};
+
+use rustc_session::parse::feature_err;
+use rustc_span::symbol::sym;
+
+crate const CHECK_CUSTOM_CODE_CLASSES: Pass = Pass {
+    name: "check-custom-code-classes",
+    run: check_custom_code_classes,
+    description: "check for custom code classes while not in nightly",
+};
+
+crate fn check_custom_code_classes(krate: Crate, cx: &DocContext<'_>) -> Crate {
+    let mut coll = CustomCodeClassLinter::new(cx);
+
+    coll.fold_crate(krate)
+}
+
+struct CustomCodeClassLinter<'a, 'tcx> {
+    cx: &'a DocContext<'tcx>,
+}
+
+impl<'a, 'tcx> CustomCodeClassLinter<'a, 'tcx> {
+    fn new(cx: &'a DocContext<'tcx>) -> Self {
+        Self { cx }
+    }
+}
+
+impl<'a, 'tcx> DocFolder for CustomCodeClassLinter<'a, 'tcx> {
+    fn fold_item(&mut self, item: Item) -> Option<Item> {
+        let cx = self.cx;
+        let dox = item.attrs.collapsed_doc_value().unwrap_or_else(String::new);
+
+        look_for_custom_classes(&cx, &dox, &item);
+
+        Some(self.fold_item_recur(item))
+    }
+}
+
+struct TestsWithCustomClasses {
+    custom_classes_found: Vec<String>,
+    nb_lang_string: usize,
+}
+
+impl crate::doctest::Tester for TestsWithCustomClasses {
+    fn add_test(&mut self, _: String, config: LangString, _: usize) {
+        for added_class in config.added_classes {
+            self.custom_classes_found.push(added_class);
+        }
+        self.nb_lang_string += 1;
+    }
+}
+
+crate fn look_for_custom_classes<'tcx>(cx: &DocContext<'tcx>, dox: &str, item: &Item) {
+    let hir_id = match cx.as_local_hir_id(item.def_id) {
+        Some(hir_id) => hir_id,
+        None => {
+            // If non-local, no need to check anything.
+            return;
+        }
+    };
+
+    let mut tests = TestsWithCustomClasses { custom_classes_found: vec![], nb_lang_string: 0 };
+
+    find_testable_code(&dox, &mut tests, ErrorCodes::No, false, None);
+
+    if !tests.custom_classes_found.is_empty() && !cx.tcx.features().custom_code_classes_in_docs {
+        debug!("reporting error for {:?} (hid_id={:?})", item, hir_id);
+        let sp = span_of_attrs(&item.attrs).unwrap_or(item.source.span());
+        feature_err(
+            &cx.tcx.sess.parse_sess,
+            sym::custom_code_classes_in_docs,
+            sp,
+            // This will list the wrong items to make them more easily searchable.
+            // To ensure the most correct hits, it adds back the 'class:' that was stripped.
+            &format!(
+                "Found in {} block(s): class:{}",
+                tests.nb_lang_string,
+                tests.custom_classes_found.join(", class:")
+            ),
```

Comment:
I find this error pretty confusing. What about this instead?
```suggestion
            "custom classes in code blocks are unstable",
       ).note(
            // This will list the wrong items to make them more easily searchable.
            // To ensure the most correct hits, it adds back the 'class:' that was stripped.
            &format!(
                "found in {} block(s): class:{}",
                tests.nb_lang_string,
                tests.custom_classes_found.join(", class:")
            ),
```

Reply:
I don't think we should list how many we found – we're pointing using the span. I think the error should say something like `found custom class on code block: class:some-custom-class`.